src/operations/blas2/trsv.hpp

```diff
@@ -0,0 +1,138 @@
+                        typename Trsv<lhs_t, matrix_t, vector_t, local_range,
+                                      is_upper, is_transposed,
+                                      is_unitdiag>::index_t &_blk_id,
+                        vector_t &_vector)
+    : lhs_(_l), matrix_(_matrix), vector_(_vector), blk_id_(_blk_id) {}
+
+template <typename lhs_t, typename matrix_t, typename vector_t,
+          uint32_t local_range, bool is_upper, bool is_transposed,
+          bool is_unitdiag>
+SYCL_BLAS_INLINE typename Trsv<lhs_t, matrix_t, vector_t, local_range, is_upper,
+                               is_transposed, is_unitdiag>::index_t
+Trsv<lhs_t, matrix_t, vector_t, local_range, is_upper, is_transposed,
+     is_unitdiag>::get_size() const {
+  return matrix_.get_size();
+}
+template <typename lhs_t, typename matrix_t, typename vector_t,
+          uint32_t local_range, bool is_upper, bool is_transposed,
+          bool is_unitdiag>
+SYCL_BLAS_INLINE bool
+Trsv<lhs_t, matrix_t, vector_t, local_range, is_upper, is_transposed,
+     is_unitdiag>::valid_thread(cl::sycl::nd_item<1> ndItem) const {
+  // Valid threads are established by ::eval.
+  return true;
+}
+
+template <typename lhs_t, typename matrix_t, typename vector_t,
+          uint32_t local_range, bool is_upper, bool is_transposed,
+          bool is_unitdiag>
+template <typename local_memory_t>
+SYCL_BLAS_INLINE typename Trsv<lhs_t, matrix_t, vector_t, local_range, is_upper,
+                               is_transposed, is_unitdiag>::value_t
+Trsv<lhs_t, matrix_t, vector_t, local_range, is_upper, is_transposed,
+     is_unitdiag>::eval(local_memory_t local_mem, cl::sycl::nd_item<1> ndItem) {
+  const index_t _offset = blk_id_ * local_range;
+
+  const index_t g_idx = _offset + ndItem.get_local_id(0);
+  const index_t l_idx = ndItem.get_local_id(0);
+
+  const index_t _N = lhs_.get_size();
+  auto l_x = local_mem.localAcc;
+
+  // copy lhs_ local memory + sync thread
+  if (g_idx < _N) l_x[l_idx] = lhs_.eval(g_idx);
+
+  ndItem.barrier(cl::sycl::access::fence_space::global_and_local);
+
+  constexpr bool is_forward =
+      (is_upper && is_transposed) || (!is_upper && !is_transposed);
+
+  const index_t n_it =
+      (_offset + local_range < _N) ? local_range : _N - _offset;
+  for (index_t _it = 0; _it < n_it; ++_it) {
+    const index_t l_diag = (is_forward) ? _it : n_it - 1 - _it;
+    const index_t g_diag = _offset + l_diag;
+
+    if (!is_unitdiag && !l_idx) l_x[l_diag] /= matrix_.eval(g_diag, g_diag);
+
+    ndItem.barrier(cl::sycl::access::fence_space::local_space);
+
+    if (((g_idx > g_diag) && (g_idx < _N) && is_forward) ||
+        ((g_idx < g_diag) && !is_forward)) {
+      const value_t val = (is_transposed) ? matrix_.eval(g_diag, g_idx)
+                                          : matrix_.eval(g_idx, g_diag);
+      l_x[l_idx] -= val * l_x[l_diag];
+    }
+
+    ndItem.barrier(cl::sycl::access::fence_space::local_space);
+  }
```

Comment:
So if `l_diag` does not meet `l_idx` I think you should be able to reduce this 3 barriers to 1 by doing something like this: 

```suggestion
  if (g_idx < _N) l_x[l_idx] = lhs_.eval(g_idx);
  constexpr bool is_forward =
      (is_upper && is_transposed) || (!is_upper && !is_transposed);

  const index_t n_it =
      (_offset + local_range < _N) ? local_range : _N - _offset;
  for (index_t _it = 0; _it < n_it; ++_it) {
    const index_t l_diag = (is_forward) ? _it : n_it - 1 - _it;
    const index_t g_diag = _offset + l_diag;

    ndItem.barrier(cl::sycl::access::fence_space::local_space);
    auto tmp = scalar(0);
    if (!is_unitdiag && !l_idx)  tmp  = l_x[l_diag] / matrix_.eval(g_diag, g_diag);

    if (((g_idx > g_diag) && (g_idx < _N) && is_forward) ||
        ((g_idx < g_diag) && !is_forward)) {
      const value_t val = (is_transposed) ? matrix_.eval(g_diag, g_idx)
                                          : matrix_.eval(g_idx, g_diag);
      l_x[l_idx] -= val * tmp
    }
      if (!is_unitdiag && !l_idx)  l_x[l_diag]=tmp;
  }
```

Reply:
Thanks for your review, @mehdi-goli!

The last commit adds the necessary modifications for removing two barriers.

Let me know if it is fine!